src/ast/nodes/shared/FunctionNode.ts

```diff
@@ -45,20 +42,14 @@ export default class FunctionNode extends NodeBase {
 		if (path.length <= 1) {
 			return false;
 		}
-		if (path[0] === 'prototype') {
-			return path.length > 2;
-		}
-		return true;
+		return path.length > 2 || path[0] !== 'prototype' || this.isPrototypeReassigned;
```

Comment:
This is about the not-so-deep reassignments. More specifically, this very important situation that is often part of initialization code is not a side-effect:
```javascript
function X(){};
X.prototype.myMethod = function() { console.log('hello'); }
```

while this is:
```javascript
function X(){};
X.doesNotExist.myMethod = function() { console.log('hello'); }
```

Note that if `X` is included for whatever reason, all prototype assignments will be included as well. I think as long as we do not have proper flow tracking, a simple flag to take care of the prototype should be enough to handle situations such as this:
```javascript
function X(){}
X.prototype.myMethod = function() {}

function Y(){}
Y.prototype = X.prototype; // this will mark the prototype of `Y` as reassigned
Y.prototype.myMethod = function() { console.log('hello'); } //  this mutation needs to be retained

const x = new X();
x.myMethod();
```

Reply:
Thanks for the clear explanation that makes complete sense. To continue to help aid my understanding I added a further question about more generic path reassignment checking below, which seems like it is related here.